python/ray/data/_internal/execution/operators/actor_pool_map_operator.py

```diff
@@ -215,6 +261,128 @@ def submit(
         yield from _map_task(fn, *blocks)
 
 
+# TODO(Clark): Promote this to a public config once we deprecate the legacy compute
+# strategies.
+@dataclass
+class AutoscalingConfig:
+    """Configuration for an autoscaling actor pool."""
+
+    # Minimum number of workers in the actor pool.
+    min_workers: int
+    # Maximum number of workers in the actor pool.
+    max_workers: int
+    # Maximum number of tasks that can be in flight for a single worker.
+    # TODO(Clark): Have this informed by the prefetch_batches configuration, once async
+    # prefetching has been ported to this new actor pool.
+    max_tasks_in_flight: int = 2
+    # Minimum ratio of ready workers to the total number of workers. If the pool is
+    # above this ratio, it will be allowed to be scaled up.
+    ready_to_total_workers_ratio: float = 0.8
+    # Maximum ratio of idle workers to the total number of workers. If the pool goes
+    # above this ratio, the pool will be scaled down.
+    idle_to_total_workers_ratio: float = 0.5
```

Comment:
Curious why we have different ratio between scale up and scale down?

Reply:
They're pretty much orthogonal! `ready` includes both active and idle workers (only pending is excluded), and `idle` only includes idle workers (both active and pending are excluded), so they're not inverses of each other. A bit more on each of these heuristics...

The 80% cap for the ready/total ratio ensures that, in the worst case, we don't try to create more actors than 125% of the cluster's CPU/GPU resource capacity; this is a basic heuristic for resource-based backpressure. This cap will probably be replaced with the executor's resource management + a heuristic based on the work queue size and misc. profiling (queue arrival rate, actor startup time, task execution time).

The 50% cap for idle/total is a super basic "more than half of the actors are idle" heuristic for when we've created too many actors; this was the simplest heuristic I could come up with that bootstraps, scales, and degrades pretty gracefully:
1. At operator startup, pending actors aren't penalized, so we're able to start an initial wave of actors in parallel.
2. The pool's steady-state should be > half of the actors are active with < half idle, so we should have solid load-balancing + an idle pool ready for input queue bursts. With load-balancing based on tasks-in-flight, this is probably a bit conservative, actually; we could look at decreasing this to 0.25 on some benchmarks to see if we really need that much pool slack.
3. As a pool becomes temporarily less active, e.g. there's a delay bubble in upstream processing, we won't downscale.

I'm guessing that we might keep this heuristic around as a super conservative idle cap, since it's a simple alternative to the common solution of idle timeouts and can be complementary (a fallback) to downscaling triggered by the executor control loop based on its resource management.

Should be fun to experiment with tweaking these going forward!

Reply:
Thanks @clarkzinzow for the detailed explanation!